frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -751,7 +778,10 @@ pub mod pallet {
                 staker_info.latest_staked_value(),
             );
 
-            if should_restake_reward {
+            let beneficiary = RewardsBeneficiary::<T>::get(&staker, &contract_id);
+            let mut weight_info = T::WeightInfo::claim_staker_without_restake();
+
+            if beneficiary.is_none() && should_restake_reward {
```

Comment:
I'm not fond of argument bloat, but perhaps the added check would be better suited inside `fn should_restake_reward`?

Reply:
Sounds good!